src/benchmarks/gc/GC.Infrastructure/Notebooks/Reports-Impl.dib

```diff
@@ -0,0 +1,1756 @@
+        {
+            //Console.WriteLine($"Exception processing {Title}");
+            //Console.WriteLine($"   {e}");
+            value = default;
+        }
+        return value;
+    }
+}
+
+public class Metric<TSource> : BaseMetric<TSource, double?>
+{
+    public string Unit;
+    public double? Cap;
+    private int _capExceededCount;
+    private double _capExceededMin;
+    private double _capExceededMax;
+    public double? AxisCountOffset;
+
+    public Metric(Func<TSource, double> extract, string title, string unit, double? cap = null, double? axisCountOffset = null)
+        : base((s => extract(s)), title)
+    {
+        Unit = unit;
+        Cap = cap;
+        AxisCountOffset = axisCountOffset;
+    }
+
+    public Metric(Func<TSource, double?> extract, string title, string unit, double? cap = null, double? axisCountOffset = null)
+        : base(extract, title)
+    {
+        Unit = unit;
+        Cap = cap;
+        AxisCountOffset = axisCountOffset;
+    }
+
+    public double? DoExtract(TSource gc, int count)
+    {
+        double? value = base.DoExtract(gc);
+        if (value.HasValue)
+        {
+            if (value > Cap)
+            {
+                _capExceededCount++;
+                _capExceededMin = Math.Min(_capExceededMin, value.Value);
+                _capExceededMax = Math.Max(_capExceededMax, value.Value);
+                value = Cap;
+            }
+            if (AxisCountOffset.HasValue) value += AxisCountOffset * count;
+        }
+        return value;
+    }
+
+    private Metric<TSource> Copy() => new(ExtractFunc, Title, Unit, Cap);
+    public Metric<TSource> WithCap(double cap) => new(ExtractFunc, Title, Unit, cap, AxisCountOffset);
+    public Metric<TSource> WithOffset(double offset) => new(ExtractFunc, Title, Unit, Cap, offset);
+
+    public void ResetDiagnostics()
+    {
+        _capExceededCount = 0;
+        _capExceededMin = double.MaxValue;
+        _capExceededMax = double.MinValue;
+    }
+
+    public void DisplayDiagnostics(string context)
+    {
+        if (_capExceededCount > 0)
+        {
+            Console.WriteLine($"Cap ({Cap.Value}) exceeded {_capExceededCount} times (min={_capExceededMin:N2}, max={_capExceededMax:N2}) for {context}");
+        }
+    }
+
+    public static Metric<TSource> Promote<TOldSource>(Metric<TOldSource> metric, Func<TSource, IEnumerable<TOldSource>> oldExtract, Aggregation aggregation)
+        => new(extract: source => aggregation.Func(oldExtract(source).Select(metric.ExtractFunc).Where(NotNull).Select(value => value.Value)),
+            title: $"{aggregation.Title} of {metric.Title}",
+            unit: aggregation.UnitOverride ?? metric.Unit);
+}
+
+public static class Metrics
+{
+    public static Metric<IterationData> Promote(Metric<TraceGC> metric, Aggregation aggregation)
+        => Metric<IterationData>.Promote(metric, iterationData => (IEnumerable<TraceGC>)iterationData.GCProcessData.GCs, aggregation);
```

Comment:
I don't know why the explicit `(IEnumerable<TraceGC>)` is required for the factored version to build.

Reply:
fixed by keeping all imports/usings together